### PR TITLE
Clear suppressed notifications only after the TimePeriod is resumed

### LINKED ISSUE #10575: Recovery notification outside time period still lost
Hello Icinga-Team :)

## Describe the bug

We just stumbled upon this issue again, though it should have been fixed with https://github.com/Icinga/icinga2/pull/10187

We have a check that recovered while outside it's notification period (04.10. 23:58) and then should have sent out the recovery notification on 06.10. 08:00.

**Time period:**
```
object TimePeriod "8x5" {
    import "legacy-timeperiod"
    ranges = {
        "friday"	= "08:00-17:00"
        "monday"	= "08:00-17:00"
        "thursday"	= "08:00-17:00"
        "tuesday"	= "08:00-17:00"
        "wednesday"	= "08:00-17:00"
    }
}
```
**Service history:**
Image

User object:
```
object User "ticketsystem" {
    import "generic-user-template-problem_down-critical"

    states = [ Critical, Down, OK, Up ]
    types = [ Custom, Problem, Recovery ]
    vars.fallback_email = "..."
}
```

Until I discovered the mentioned issue I wasn't sure if this is "works as designed", but apparently it shouldn't be.

This is the debug log from around the time the notification was suppressed
```
[2025-10-03 23:58:24 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-03 23:58:24 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!pending_updates!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-03 23:58:25 +0200] debug/ApiListener: Not checking connection to Zone 'abcd-dc02' because it's not in the same zone, a parent or a child zone.
[2025-10-03 23:58:25 +0200] debug/ApiListener: Not checking connection to Zone 'abcd-dc02' because it's not in the same zone, a parent or a child zone.
[2025-10-03 23:58:25 +0200] debug/ApiListener: Not checking connection to Zone 'abcd-dc02' because it's not in the same zone, a parent or a child zone.
[2025-10-03 23:58:27 +0200] debug/ApiListener: Not checking connection to Zone 'abcd-dc02' because it's not in the same zone, a parent or a child zone.
[2025-10-03 23:58:29 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-03 23:58:29 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!pending_updates!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-03 23:58:29 +0200] information/Checkable: Acknowledgement cleared for checkable 'abcd-dc02!commvault-backup'.
[2025-10-03 23:58:29 +0200] notice/ApiListener: Received config delete for object: {"name":"abcd-dc02!commvault-backup!37e51972-af95-480d-b266-faf9c7be6fa2","type":"Comment","version":1759471212.819265}
[2025-10-03 23:58:29 +0200] notice/ApiListener: Discarding 'config delete object' message from 'abcd-lm1-fra' (endpoint: 'abcd-lm1-fra', zone: 'abcd-fra') for object 'abcd-dc02!commvault-backup!37e51972-af95-480d-b266-faf9c7be6fa2' of type 'Comment'. Sender is in a child zone.
[2025-10-03 23:58:29 +0200] notice/Comment: Removed comment 'abcd-dc02!commvault-backup!37e51972-af95-480d-b266-faf9c7be6fa2' from object 'abcd-dc02!commvault-backup'.
[2025-10-03 23:58:29 +0200] information/ConfigObjectUtility: Deleted object 'abcd-dc02!commvault-backup!37e51972-af95-480d-b266-faf9c7be6fa2' of type 'Comment'.
[2025-10-03 23:58:29 +0200] notice/Checkable: State Change: Checkable 'abcd-dc02!commvault-backup' hard state change from CRITICAL to OK detected.
[2025-10-03 23:58:29 +0200] information/Checkable: Checkable 'abcd-dc02!commvault-backup' has 1 notification(s). Checking filters for type 'Recovery', sends will be logged.
[2025-10-03 23:58:29 +0200] notice/Notification: Attempting to send notifications of type 'Recovery' for notification object 'abcd-dc02!commvault-backup!notification-jira-service_critical'.
[2025-10-03 23:58:29 +0200] notice/Notification: Not sending notifications for notification object 'abcd-dc02!commvault-backup!notification-jira-service_critical': not in timeperiod '8x5'
```

and here the log from when the notification was expected
```
[2025-10-06 07:59:32 +0200] debug/ScheduledDowntime: Creating new Downtime for ScheduledDowntime "abcd-dc02!commvault-logbackup!Commvault-Maintenance Services"
[2025-10-06 07:59:32 +0200] debug/ScheduledDowntime: Creating new Downtime for ScheduledDowntime "abcd-dc02!commvault-backup!Commvault-Maintenance Services"
[2025-10-06 07:59:37 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-06 07:59:42 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-06 07:59:47 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-06 07:59:52 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-06 07:59:58 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-06 08:00:03 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-06 08:00:09 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-06 08:00:14 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-06 08:00:19 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-06 08:00:24 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-06 08:00:29 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-06 08:00:35 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
[2025-10-06 08:00:35 +0200] debug/ScheduledDowntime: Creating new Downtime for ScheduledDowntime "abcd-dc02!commvault-logbackup!Commvault-Maintenance Services"
[2025-10-06 08:00:35 +0200] debug/ScheduledDowntime: Creating new Downtime for ScheduledDowntime "abcd-dc02!commvault-backup!Commvault-Maintenance Services"
[2025-10-06 08:00:52 +0200] notice/ApiListener: Received config update for object: {"config":"object Downtime \"85f19ab7-7ef1-4ce7-bd6c-eb2302e328f0\" ignore_on_error {\n\tauthor = \"Icinga Admin DSL\"\n\tauthoritative_zone = \"master\"\n\tcomment = \"Scheduled Update\"\n\tconfig_owner = \"abcd-dc02!commvault-backup!scheduled_update\"\n\tconfig_owner_hash = \"1cea808495b94018c496a5b71c0ff4d6202e717eb4017f5e38c9229287418dcd\"\n\tduration = 3600.000000\n\tend_time = 1760904000.000000\n\tentry_time = 1759269882.441912\n\tfixed = false\n\thost_name = \"abcd-dc02\"\n\tparent = \"\"\n\tscheduled_by = \"abcd-dc02!commvault-backup!scheduled_update\"\n\tservice_name = \"commvault-backup\"\n\tstart_time = 1760896800.000000\n\ttriggered_by = \"\"\n\tversion = 1759269882.441952\n\tzone = \"abcd-fra\"\n}\n","modified_attributes":{},"name":"abcd-dc02!commvault-backup!85f19ab7-7ef1-4ce7-bd6c-eb2302e328f0","original_attributes":[],"type":"Downtime","version":1759269882.441952,"zone":"abcd-fra"}
[2025-10-06 08:00:52 +0200] notice/ApiListener: Discarding 'config update object' message from 'abcd-lm1-fra' (endpoint: 'abcd-lm1-fra', zone: 'abcd-fra') for object 'abcd-dc02!commvault-backup!85f19ab7-7ef1-4ce7-bd6c-eb2302e328f0' of type 'Downtime'. Sender is in a child zone.
[2025-10-06 08:00:52 +0200] notice/ApiListener: Received config update for object: {"config":"object Downtime \"52e32f9d-e102-4a4c-9048-1ea3e2535548\" ignore_on_error {\n\tauthor = \"Icinga Admin DSL\"\n\tauthoritative_zone = \"master\"\n\tcomment = \"Scheduled Update\"\n\tconfig_owner = \"abcd-dc02!commvault-logbackup!scheduled_update\"\n\tconfig_owner_hash = \"1cea808495b94018c496a5b71c0ff4d6202e717eb4017f5e38c9229287418dcd\"\n\tduration = 3600.000000\n\tend_time = 1760904000.000000\n\tentry_time = 1759269882.342413\n\tfixed = false\n\thost_name = \"abcd-dc02\"\n\tparent = \"\"\n\tscheduled_by = \"abcd-dc02!commvault-logbackup!scheduled_update\"\n\tservice_name = \"commvault-logbackup\"\n\tstart_time = 1760896800.000000\n\ttriggered_by = \"\"\n\tversion = 1759269882.342464\n\tzone = \"abcd-fra\"\n}\n","modified_attributes":{},"name":"abcd-dc02!commvault-logbackup!52e32f9d-e102-4a4c-9048-1ea3e2535548","original_attributes":[],"type":"Downtime","version":1759269882.342464,"zone":"abcd-fra"}
[2025-10-06 08:00:52 +0200] notice/ApiListener: Discarding 'config update object' message from 'abcd-lm1-fra' (endpoint: 'abcd-lm1-fra', zone: 'abcd-fra') for object 'abcd-dc02!commvault-logbackup!52e32f9d-e102-4a4c-9048-1ea3e2535548' of type 'Downtime'. Sender is in a child zone.
[2025-10-06 08:00:55 +0200] notice/NotificationComponent: Reminder notification 'abcd-dc02!commvault-backup!notification-jira-service_critical': Notification was sent out once and interval=0 disables reminder notifications.
```

Not sure about the `Creating Downtime` messages, because that is an old ScheduledDowntime object with an specific date
```
apply ScheduledDowntime "Commvault-Maintenance Services" to Service {
    author = "abcd"
    comment = "Downtime für Commvault-Updates oder andere Commvault Wartungen.
    fixed = true
    assign where match("commvault*", service.name)
    ranges = {
        "2025-08-25"	= "08:00-17:00"
    }
}
```

But as the screenshot shows there was no downtime entered and imo the notification should have been sent out at 08:00 on 06.10.2025

The question is, why wasn't the notification triggered?

## Your Environment

Include as many relevant details about the environment you experienced the problem in

* Version used (`icinga2 --version`): 2.15.0
* Operating System and version: rhel8.10
* Single master instance with 90+ connected zones.
* Host that the notification is missing for belongs to a satellite zone.

Best regards

### PR DESCRIPTION
**Pull request by jschmidt-icinga**
### Bug Description :lady_beetle: 

When [multiple state changes](https://github.com/Icinga/icinga2/issues/10575#issuecomment-3422277439) happen outside a `TimePeriod`, `NotificationComponent::FireSuppressedNotifications()` will clear notifications that *don't apply* too soon, leading the state of the suppressed notifications to become inconsistent. 

https://github.com/Icinga/icinga2/blob/5d46ca4f770f8b013c38400bc61af8388437e712/lib/notification/notificationcomponent.cpp#L86-L91

Ultimately this will cause any suppressed notifications to be discarded when the `TimePeriod` resumes.

### Fix Description :wrench: 

The [trivial fix](https://github.com/Icinga/icinga2/commit/ee1fb149377683dab221840509a7fcc96f157d7b) is to just move the check on `NotificationReasonApplies()` inside the `tp->IsInside()` condition, thereby removing them after the `TimePeriod` resumes and leaving the list of suppressed notifications intact until no further notifications will be added to it.

I'm not entirely certain what the purpose of that check is, but the only scenario I could think of is when a `TimePeriod` begins and a check result produces a notification before `NotificationTimerHandler` can run and send out the suppressed notifications. I've added a unit-test to ensure that behavior is the same as before.

@Al2Klimov, since you're the one who wrote the original code, maybe you can give other scenarios where this check might be affecting and that could potentially stop working with this PR. In that case we can add additional test-cases and look for a more complex fix.

### Unit-Tests :white_check_mark: 

I've added tests for some common scenarios including the bugged behavior of the issue this PR fixes. This way the behavior of the `NotificationComponent` can be verified before and after this PR. I'm happy to add additional test-cases for behavior that might be affected by this issue and the fix.

Fixes #10575

### CODE REVIEW

File: lib/notification/notificationcomponent.cpp [hunk 1]
```diff
@@ -98,6 +91,11 @@ void FireSuppressedNotifications(const Notification::Ptr& notification)
 				if (!(suppressedTypes & type) || checkable->NotificationReasonSuppressed(type))
 					continue;
 
+				if ((suppressedTypes & type) && !checkable->NotificationReasonApplies(type)) {
+					subtract |= type;
+					suppressedTypes &= ~type;
```
Comment (Al2Klimov) [inline lib/notification/notificationcomponent.cpp]: Didn't read the issue, but the fix shouldn't hurt. After all, this is also the code branch where notifications are sent.

File: lib/notification/notificationcomponent.cpp [hunk 2]
```diff
@@ -81,23 +81,20 @@ void FireSuppressedNotifications(const Notification::Ptr& notification)
...
 	if (suppressedTypes) {
 		auto tp (notification->GetPeriod());
+		auto checkable (notification->GetCheckable());
 
 		if ((!tp || tp->IsInside(Utility::GetTime())) && !checkable->IsLikelyToBeCheckedSoon()) {
 			for (auto type : {NotificationProblem, NotificationRecovery, NotificationFlappingStart, NotificationFlappingEnd}) {
 				if (!(suppressedTypes & type) || checkable->NotificationReasonSuppressed(type))
 					continue;
 
+				if ((suppressedTypes & type) && !checkable->NotificationReasonApplies(type)) {
```
Comment (yhabteab) [inline lib/notification/notificationcomponent.cpp]: The first check of this condition is superfluous and can be removed.

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -129,6 +130,7 @@ set(base_test_SOURCES
   $<TARGET_OBJECTS:remote>
   $<TARGET_OBJECTS:icinga>
   $<TARGET_OBJECTS:methods>
+  $<TARGET_OBJECTS:notification>
```
Comment (yhabteab) [inline test/CMakeLists.txt]: Please add these dependencies conditionally based on the `ICINGA2_WITH_NOTIFICATION` variable.

File: test/notification-notificationcomponent.cpp [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+#include "base-testloggerfixture.hpp"
+#include "config/configcompiler.hpp"
+#include "notification/notificationcomponent.hpp"
+
+using namespace icinga;
+
+namespace {
+
+// black magic
+template<auto prvPtr>
```
Comment (yhabteab) [inline test/notification-notificationcomponent.cpp]: That's indeed a black magic, but describing it a bit more what it does and it's used for would help readers. Also, does the abbreviated `prv` of the template param name stand for `private` or `previous`? Maybe use a bit of descriptive name instead.

File: test/notification-notificationcomponent.cpp [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+struct InvokeTimerHandlerImpl
+{
+	friend void InvokeTimerHandler(const NotificationComponent::Ptr& nc) { return ((*nc).*prvPtr)(); }
```
Comment (yhabteab) [inline test/notification-notificationcomponent.cpp]:
```suggestion
	friend void InvokeTimerHandler(const NotificationComponent::Ptr& nc) { return (*nc.*prvPtr)(); }
```

File: test/notification-notificationcomponent.cpp [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+		m_TimePeriod = TimePeriod::GetByName("tp1");
+		BOOST_REQUIRE(m_TimePeriod);
+
+		ApiListener::UpdateObjectAuthority();
+		BOOST_REQUIRE(ApiListener::UpdatedObjectAuthority());
+
+		// Store the old periods from the config snippets to reuse them later.
+		m_AllTimePeriod = m_TimePeriod->GetRanges();
+
+		Service::OnNotificationSentToUser.connect(NotificationSentToUserHandler);
```
Comment (yhabteab) [inline test/notification-notificationcomponent.cpp]: Using service to register the signal here looks odd! Why not just use `Checkable` instead?

File: test/notification-notificationcomponent.cpp [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+	SetNotificationInverval(0.15);
+	ReceiveCheckResults(2, ServiceCritical);
+
+	// The first run of the timer sets up the next reminder notification.
+	NotificationTimerHandler();
+	BOOST_REQUIRE(WaitForExpectedNotificationCount(1));
+	BOOST_REQUIRE_EQUAL(GetLastNotification(), NotificationProblem);
+
+	// Rerunning the timer before the next interval should not trigger a reminder notification.
+	NotificationTimerHandler();
```
Comment (yhabteab) [inline test/notification-notificationcomponent.cpp:228]: Isn't the interval set to `0.15` at the beginning of this test case? How can you be sure that it's going to call this function before the `15ms` interval has passed? Apart from that, the `WaitForExpectedNotificationCount` call above can block `5ms` in the worst case, so I don't get how this all fit with what this comment says.

If I understand what this test is supposed to achieve, then I would just call `NotificationTimerHandler()` repeatedly before performing any assertions above instead, so you only get a very tight time window for race conditions and these assertions can be merged with the above ones.

**Review by yhabteab — COMMENTED** [review]
> I'm not entirely certain what the purpose of that check is, but the only scenario I could think of is when a `TimePeriod` begins and a check result produces a notification before `NotificationTimerHandler` can run and send out the suppressed notifications.

That's pretty much the reasoning behind it. Its purpose is to avoid sending notifications that are no longer applicable, e.g. a problem notification when the checkable has already recovered. For instance, if a problem notification was suppressed due to being outside the time period, but recovers right after the time period starts and before the `NotificationTimerHandler` runs, it won't clear the suppressed problem notification, thus sending out the suppressed problem notification would be incorrect.

File: lib/notification/notificationcomponent.cpp (hunk 2 above)
Comment (yhabteab) [reply lib/notification/notificationcomponent.cpp]: Oh, and the whole `if (suppressedTypes) ...` surrounding this loop has also become redundant, thus you can either remove the whole `if` or the other `if (!suppressedTypes) ...` at the beginning of this function. Your choice.

File: test/notification-notificationcomponent.cpp (hunk 7 above)
Comment (jschmidt-icinga) [reply test/notification-notificationcomponent.cpp:228]: >How can you be sure that it's going to call this function before the 15ms interval has passed?

Firstly, it's 150ms, not 15ms and `WaitForExpectedNotificationCount()` has a timeout of 5s, not 5ms, but that's more so failing tests (like when not notification is sent at all) don't block the runners forever. The "wait" is just needed because sending notifications goes through an async-queue which has a minimal latency, but that shouldn't be more than a few ms even on complete potato-hardware. The other test-cases, most of them with multiple calls to that function, usually complete in less than 100ms total. It seems to be very unpredictable on the Windows runners though, which is why I increased it from 50ms to 150ms.

I'd prefer not to merge the two groups of assertions together like you suggest, because that would make it less clear what is being tested. Maybe you can agree that under the more generous timings above it should be fine?

File: test/notification-notificationcomponent.cpp (hunk 4 above)
Comment (jschmidt-icinga) [reply test/notification-notificationcomponent.cpp]: I thought that I would leave some fun for the reader to discover on their own, since it's such a weird (but not UB!) language edge-case. But I'm realizing now that it looks like something I just copy pasted from somewhere or got as an answer from Chat-GPT without understanding what it does. I will add an explanation how this "friend injection" works and change the parameter name.

File: test/notification-notificationcomponent.cpp (hunk 6 above)
Comment (jschmidt-icinga) [reply test/notification-notificationcomponent.cpp]: I copy-pasted it from `notification.cpp`, where it was used that way, but you're right.

File: lib/notification/notificationcomponent.cpp (hunk 2 above)
Comment (jschmidt-icinga) [reply lib/notification/notificationcomponent.cpp]: I did this, but now the diff is all messed up. I tried to clean it up, but it seems Github just isn't going to show the body of the removed `if` as a whitespace change. :shrug:

File: test/notification-notificationcomponent.cpp (hunk 6 above)
Comment (yhabteab) [reply test/notification-notificationcomponent.cpp]: Forgot to change it?

File: test/notification-notificationcomponent.cpp (hunk 7 above)
Comment (yhabteab) [reply test/notification-notificationcomponent.cpp:228]: > Firstly, it's 150ms, not 15ms and `WaitForExpectedNotificationCount()` has a timeout of 5s, not 5ms,

Overseen the 🐘! 😃.

> Maybe you can agree that under the more generous timings above it should be fine?

Fine! One can also easily retry the tests if that generous timout isn't big enough on some runners sometimes.

File: lib/notification/notificationcomponent.cpp (hunk 2 above)
Comment (yhabteab) [reply lib/notification/notificationcomponent.cpp]: > but it seems Github just isn't going to show the body of the removed if as a whitespace change. 🤷

Are you sure? GitHub shows me this, when I change my settings to hide whitespace only changes.

Bildschirmfoto 2025-11-03 um 13 34 00

File: test/notification-notificationcomponent.cpp [hunk 8]
```diff
@@ -0,0 +1,501 @@
+/* Icinga 2 | (c) 2025 Icinga GmbH | GPLv2+ */
+
+#include <BoostTestTargetConfig.h>
+#include "base/defer.hpp"
+#include "remote/apilistener.hpp"
+#include "base-testloggerfixture.hpp"
```
Comment (yhabteab) [inline test/notification-notificationcomponent.cpp]: Maybe you should also prefix this include with `test/` for consistency.

File: lib/notification/notificationcomponent.cpp (hunk 2 above)
Comment (jschmidt-icinga) [reply lib/notification/notificationcomponent.cpp]: You're right. I didn't have "Hide whitespace" enabled in Github's diff view.
